Add Conditions to Ensure Bind Succeeds with `Transfer-Encoding: chunked`

178inaba opened the pull request:
In Go, when the length of the Body is unknown, `ContentLength` is set to -1. As demonstrated in #2716, this applies to cases where `Transfer-Encoding: chunked` is used.

```
	// ContentLength records the length of the associated content.
	// The value -1 indicates that the length is unknown.
	// Values >= 0 indicate that the given number of bytes may
	// be read from Body.
	//
	// For client requests, a value of 0 with a non-nil Body is
	// also treated as unknown.
	ContentLength [int64](https://pkg.go.dev/builtin#int64)
```

https://pkg.go.dev/net/http#Request

Previously, only `0` was excluded during Bind, but starting from #2710, `-1` was also excluded, making it impossible to Bind requests with `Transfer-Encoding: chunked`. I have fixed this issue.


Fixes #2716.

aldas commented on bind.go:
```diff
@@ -67,7 +67,14 @@ func (b *DefaultBinder) BindQueryParams(c Context, i interface{}) error {
 // See MIMEMultipartForm: https://golang.org/pkg/net/http/#Request.ParseMultipartForm
 func (b *DefaultBinder) BindBody(c Context, i interface{}) (err error) {
 	req := c.Request()
-	if req.ContentLength <= 0 {
```
```go
	if req.ContentLength == 0 {
		return
	} else if req.ContentLength < 0 {
		var isChunked bool
		for _, enc := range req.TransferEncoding {
			if enc == "chunked" {
				isChunked = true
				break
			}
		}
		if !isChunked {
			return
		}
	}
```

would be better. we only should check chunked when we are dealing with -1

but I think it would be even better to go back to `if req.ContentLength == 0 {` check

aldas reviewed (changes requested):
I think going back to == 0 check is better.  

and lets add your test for chunked and modify that failing test to test error case and add another empty body case

```go
		{
			name:             "nok, JSON POST bind to http.NoBody",
			givenURL:         "/api/real_node/endpoint?node=xxx",
			givenMethod:      http.MethodPost,
			givenContentType: MIMEApplicationJSON,
			givenContent:     http.NoBody,
			expect:           &Node{ID: 0, Node: ""},
			expectError:      "code=400, message=EOF, internal=EOF",
		},
		{
			name:             "ok, JSON POST with empty body",
			givenURL:         "/api/real_node/endpoint?node=xxx",
			givenMethod:      http.MethodPost,
			givenContentType: MIMEApplicationJSON,
			givenContent:     strings.NewReader(``),
			expect:           &Node{ID: 0, Node: ""},
		},
		{
			name:             "ok, JSON POST bind to struct with: path + query + chunked body",
			givenURL:         "/api/real_node/endpoint?node=xxx",
			givenMethod:      http.MethodPost,
			givenContentType: MIMEApplicationJSON,
			givenContent:     httputil.NewChunkedReader(strings.NewReader("18\r\n" + `{"id": 1, "node": "zzz"}` + "\r\n0\r\n")),
			whenChunkedBody:  true,
			expect:           &Node{ID: 1, Node: "zzz"},
		},
```

178inaba commented:
@aldas 
Thank you for the review!
I reverted the ContentLength check to only 0 and added tests.
Should we add a condition for ContentLength == -1 && chunked?

aldas commented:
release [v4.13.1](https://github.com/labstack/echo/releases/tag/v4.13.1 ) is available
